# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0 late days

In [3]:
# Imports section
import pandas as pd
import numpy as np
from sklearn import svm
from sklearn import datasets
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

## 1. Loading the dataset

In [4]:
# Using pandas load the dataset
dataframe = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(dataframe.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(dataframe.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [5]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataframe[['Temperature °C', 'Mols KCL']]
Y = dataframe['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
x_train, x_test, y_train, y_test = train_test_split(X, Y, train_size=0.9, random_state=42)

## 3. Perform a Linear Regression

In [6]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(x_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = pd.DataFrame([[200, 700]])
predicted_size = model.predict(sample_data)
print(predicted_size)

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print(model.score(x_train, y_train))

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print("Coefficients:", coefficients)
print("Intercept:", intercept)

[486724.34963103]
0.8610082946223788
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


c:\Users\saran\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$) 

$h(x) = 866.14641337 \cdot T + 1032.69506649 \cdot mols - -409391.47958340764$

Report on score and explain meaning: This score is suggesting the model is working just fine, because the scale is from 0 to 1 and the closer to 1 resembles the better the performance and 0.86 is quite close to 1. This model is achieving 86.1% of the best performance which is great. This model still has some optimizations to be made which can be done by providing more training points.

## 4. Use Cross Validation

In [7]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cross_val = cross_val_score(model, x_train, y_train, cv=5)
print(cross_val)
print("Average:", sum(cross_val)/5)
# Report on their finding and their significance

[0.86226163 0.81982226 0.88938198 0.86663176 0.85729958]
Average: 0.8590794402339516


Write findings here: The cross-validation process resulted in an average score of 0.859, indicating that the model performs well overall. The low variability across folds, with a tight range of just 0.07, suggests that the model generalizes consistently across different subsets of the data. This consistency reduces the risk of overfitting or underfitting, leading to more reliable predictions. However, despite the strong performance, further refinements could potentially improve the model’s accuracy.

## 5. Using Polynomial Regression

In [8]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
x_train_poly = poly.fit_transform(x_train)
x_test_poly = poly.transform(x_test)
model = LinearRegression()
model.fit(x_train_poly, y_train)
sample_data = pd.DataFrame([[200, 700]])
sample_data_poly = poly.transform(sample_data)
predicted_size = model.predict(sample_data_poly)
print(predicted_size)
train_score = model.score(x_train_poly, y_train)
print(train_score)
coefficients = model.coef_
intercept = model.intercept_
print("Coefficients:", coefficients)
print("Intercept:", intercept)
# Perform k-fold cross validation (as above)
cross_val = cross_val_score(model, x_train_poly, y_train, cv=5)
print(cross_val)
# Report on the metrics and output the resultant equation as you did in Part 3.

[296399.99999713]
1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27196606e-07  1.26450728e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0478793885558844e-05
[1. 1. 1. 1. 1.]


c:\Users\saran\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names
  warnings.warn(


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$h(x) = (12.0 \cdot T + 2.0 \cdot (Mol \cdot KCL^2)) + .02857(T, mol)nm^3$

Report on the score and interpret: The train score of 1 means that the model fit our data perfectly and completely. It also verifies that the derived equation is very accurate.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [10]:
# Load the dataset. Then train and evaluate the classification models.
data = pd.read_csv('ckd_feature_subset.csv')

X = data.drop(columns=['Target_ckd']) 
y = data['Target_ckd']
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, random_state=42)

def evaluate_model(model, X, y):
    scores = cross_val_score(model, X, y, cv=5)
    return scores.mean(), scores.std()

# Initialize classifiers
models = {
    "Logistic Regression": LogisticRegression(max_iter=200),
    "Support Vector Machines": SVC(),
    "k-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(max_iter=1000)
}

# Evaluate each model
results = {}
for model_name, model in models.items():
    mean_accuracy, std_dev = evaluate_model(model, X, y)
    results[model_name] = (mean_accuracy, std_dev)

# Create a DataFrame for results
results_df = pd.DataFrame(results, index=["Mean Accuracy", "Standard Deviation"]).T
print(results_df)


                         Mean Accuracy  Standard Deviation
Logistic Regression           0.856129            0.054236
Support Vector Machines       0.928172            0.059993
k-Nearest Neighbors           0.928172            0.031893
Neural Network                0.941290            0.024040


## Results and Conclusion for Classification Experiments

Overall, each model performed pretty well with the lowest having a mean accuracy of 0.856. The Logistic Regression model performed the worst which could be because this model is linear in nature and does not capture the complexities in the data. The SVC and KNN models had the same mean accuracy but with different standard deviations. This suggests that SVC could vary across different subsets of data showing more sensitivity than the KNN model. The best model was the neural network likely due to its capacity to learn complex patterns within the features. It had the highest accuracy and lowest standard deviation indicating it is extremely accurate and robust to variations in the dataset.

In [ ]:
# Experiments with Neural Network. Report of 3 minimum
nn_configurations = {
    "Config 1": MLPClassifier(hidden_layer_sizes=(50,), max_iter=1000),
    "Config 2": MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000),
    "Config 3": MLPClassifier(hidden_layer_sizes=(50, 25), max_iter=1000)
}

nn_results = {}
for config_name, model in nn_configurations.items():
    mean_accuracy, std_dev = evaluate_model(model, X, y)
    nn_results[config_name] = (mean_accuracy, std_dev)

# Create a DataFrame for Neural Network results
nn_results_df = pd.DataFrame(nn_results, index=["Mean Accuracy", "Standard Deviation"]).T
print(nn_results_df)

          Mean Accuracy  Standard Deviation
Config 1       0.927957            0.038746
Config 2       0.941290            0.024040
Config 3       0.967312            0.029342


## Results and Conclusion for Neural Network Experiments

Overall every configuration had a very high mean accuracy suggesting how reliable neural networks are. There is a significant improvement in accuracy from Config 1 to Config 3 indicating that adding a second layer (and the additional neurons) allows the model to capture more complex patterns in the data. Increasing the number of neurons from 50 (Config 1) to 100 (Config 2) appears to positively affect the model with more consistency and accuracy. Overall, the parameter of number of layers in each configuration made the biggest difference in which it would significantly enhance the classification of the model for this dataset. Neural networks are very effective and the more parameters, the better they perform.